# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [97]:
# import numpy and pandas

import numpy as np
import pandas as pd
import scipy 
import stats
from scipy.stats import ttest_ind
from scipy.stats import ttest_rel
from scipy.stats import ttest_1samp
from scipy.stats import chi2_contingency

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [2]:
# Run this code:

pokemon = pd.read_csv('../pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [3]:
# Your code here:

pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [47]:
pokerecuento = pokemon['Legendary'].value_counts()

In [49]:
pokerecuento

False    735
True      65
Name: Legendary, dtype: int64

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [34]:
#POKEMON DOMINGUERO
std_P_dominguero = pokemon_dominguero['Total'].std()
mean_P_dominguero = pokemon_dominguero['Total'].mean()
print('La media y desviación típica de los pokémon domingueros son {} y {} respectivamente'.format(mean_P_dominguero, std_P_dominguero))

La media y desviación típica de los pokémon domingueros son 417.21360544217686 y 106.76041745713005 respectivamente


In [40]:
#POKEMON LEGENDARIO
std_P_legendario = pokemon_lengendario['Total'].std()
mean_P_legendario = pokemon_lengendario['Total'].mean()
print('La media y desviación típica de los pokémon legendarios son {} y {} respectivamente'.format(mean_P_legendario, std_P_legendario))

La media y desviación típica de los pokémon legendarios son 637.3846153846154 y 60.93738905315344 respectivamente


The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [50]:
pokemon_dominguero = pokemon[pokemon['Legendary']==False]
pokemon_lengendario = pokemon[pokemon['Legendary']==True]

ttest_ind(pokemon_dominguero.Total, pokemon_lengendario.Total, equal_var = False)

Ttest_indResult(statistic=-25.8335743895517, pvalue=9.357954335957446e-47)

What do you conclude from this test? Write your conclusions below.

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [52]:
# Your code here:
poketipos = pokemon['Type 1'].value_counts()
poketipos

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Rock         44
Electric     44
Ghost        32
Dragon       32
Ground       32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [58]:
std_P_water = pokemon[pokemon['Type 1']=='Water'].std()
mean_P_water = pokemon[pokemon['Type 1']=='Water'].mean()
print('La media y desviación típica de los pokémon de agua son {} y {} respectivamente'.format(mean_P_water.Total, std_P_water.Total))

La media y desviación típica de los pokémon de agua son 430.45535714285717 y 113.1882660643146 respectivamente


Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

I.C: 95%
α : 0.05

Ho: Son iguales
Hi: No son iguales

In [ ]:
pokemons_water = pokemon[pokemon['Type 1']=='Water']
pokemon_otherelem = pokemon[pokemon['Type 1']!='Water']

ttest_ind(pokemons_water.Total, pokemon_otherelem.Total, equal_var = False)

In [63]:
pvalue = 0.6433915385821449/2
pvalue

0.32169576929107246

Write your conclusion below.

Dado un P-Valor de 0.32169576929107246 NO rechaza la Hipótesis de que los pokémon pudieran ser iguales ya que es superior al nivel de significancia del 5%.



# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [78]:
pokeataque = pokemon['Attack']
pokedefensa = pokemon['Defense']

ttest_rel(pokeataque, pokedefensa)

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

Describe the results of the test in the cell below.

Dado un P-Valor de 1.7140303479358558e-05 se rechaza la Hipótesis de que el Ataque y la Defensa puederan ser variables dependientes ya que es inferior al nivel de significancia del 5%.



We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [ ]:
pokeataque_flipado = pokemon['Sp. Atk']
pokedefensa_flipada = pokemon['Sp. Def']

ttest_rel(pokeataque_flipado, pokedefensa_flipada)

In [80]:
pvalue=0.3933685997548122/2
pvalue

0.1966842998774061

Describe the results of the test in the cell below.

As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [87]:
ttest_1samp(pokeataque - pokedefensa, 0)

Ttest_1sampResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [91]:
poketab = pd.crosstab(pokemon['Legendary'], pokemon['Type 1']=='Water')
poketab



Type 1     False  True 
Legendary              
False        627    108
True          61      4

Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [ ]:
chi2_contingency(poketab, correction=False)

In [100]:
pvalue=0.0571762700873503/2
pvalue

0.02858813504367515

Based on a 95% confidence, should we reject the null hypothesis?